### ISSUE TITLE
Adding Eigen solution driver routines to egg.

### ISSUE THREAD

**Issue by ThatGeoGuy** [iraikov/chicken-lapack#2]
This is somewhat of a feature request to complete some of the functionality made available via LAPACK. Currently, the egg only supports optimized LAPACK routines made available via ATLAS; however, most users will compile ATLAS with the complete LAPACK package using the following `configure` flag: 

```
--with-netlib-lapack-tarfile=/path/to/netlib/lapack-x.x.x.tgz
```

where the `x`s in the file name correspond to the version of lapack compiled with ATLAS. Moreover, most Linux distributions (or package managers in general) likewise package the full lapack functionality with ATLAS. To see if you have all of the functions, look for `dgeev_` or another driver routine that is in LAPACK but not provided via ATLAS in the symbols table:

```
$ nm /usr/lib/liblapack.a | grep dgeev
dgeev.o:
0000000000000000 T dgeev_
dgeevx.o:
0000000000000000 T dgeevx_
```

This means that these functions are available, but not fully provided via the ATLAS interface. You can compile a regular C program to take advantage of these functions as follows: 

```
// Example C program
#include <stdio.h>
#include <stdlib.h>
#include <clapack.h>

int main() {
    dgeev_( /* args here */ );
    return 0;
}
```

Which is then compiled with: 

```
$ gcc example.c -o example -llapack -lf77blas -lgfortran -lcblas -latlas -lm
```

So  theoretically, this means that by perhaps adding `libf77blas` and `libgfortran` to the compiler flags in the .setup file, as well as making some changes to expose the interface to dgeev_ via CHICKEN scheme's bind egg, it should be plausible to expose additional LAPACK routines from this egg. The reason in particular I chose dgeev_ is because I have a vested interest in Eigenvalues / Eigenvectors, and I think beyond the general matrix solvers and the symmetric-positive-(semi)-definite solvers in LAPACK / ATLAS the Eigen solution functions are the most useful and versatile. 

I don't yet know what changes to make to expose dgeev_ to scheme bindings, but I believe it will takes some changes to existing code, because currently the code works by prepending `clapack_` to any of the called bindings (e.g. if you call `dgesv` in scheme it will expand and call to `clapack_dgesv` in ATLAS). This ought to be a good start to figuring out  the problem in any case, any suggestions or advice are appreciated, as I am willing to help implement this feature.

**Comment by iraikov** [comment]
One possibility is to create a second module for the lapack bindings not covered by atlas. Then you wouldn't need to worry about prepending clapack_ to the functions, and lapack-wrap\* macros can be reused with minimal modifications.

**Comment by ThatGeoGuy** [comment]
Hey @iraikov, do you mean to define a second module within this egg? I would be hesitant to define LAPACK routines as a separate module outside this egg, if only because it would fragment the library functions. 

I think the biggest hurdle is the lapack-wrap macro itself, but as a separate module, most of the problems go away. Something like atlas-lapack-extras or atlas-lapack-other might be a good name for such a module. Similar to how CHICKEN implements srfi-41 as streams, streams-derived, and streams-utils. In any case, I'll work on the idea, and once I have something more tangible I'll submit a pull request.

**Comment by iraikov** [comment]
Yes, I meant a second module within the egg. Why don't you try modifying
the lapack-wrap macro in a separate module? lapack-extras or lapack-other
sounds
good to me.

**Comment by ThatGeoGuy** [comment]
Hey @iraikov, I've modified some parts of the lapack-wrap macro, and more or less got the `geev_` functions compiling. You can see the changes here: https://github.com/ThatGeoGuy/chicken-lapack/tree/lapack-extras

Now, my only major problem is testing the functions, to see if I've managed to get everything right. I haven't quite figured out how lapack-wrapx wraps errors that occur, and I'm having some trouble calling the functions, especially since they take pointers to everything, including arguments that should just be integers. Could you advise me on how I could go about testing the function's interface? For reference, I'll give you the documentation I have on dgeev_:

``` C
int dgeev_(char *jobvl, char *jobvr, integer *n, doublereal *
    a, integer *lda, doublereal *wr, doublereal *wi, doublereal *vl, 
    integer *ldvl, doublereal *vr, integer *ldvr, doublereal *work, 
    integer *lwork, integer *info);
```

And the argument documentation:

```
Purpose   
=======   

DGEEV computes for an N-by-N real nonsymmetric matrix A, the   
eigenvalues and, optionally, the left and/or right eigenvectors.   

The right eigenvector v(j) of A satisfies   
                 A * v(j) = lambda(j) * v(j)   
where lambda(j) is its eigenvalue.   
The left eigenvector u(j) of A satisfies   
              u(j)**H * A = lambda(j) * u(j)**H   
where u(j)**H denotes the conjugate transpose of u(j).   

The computed eigenvectors are normalized to have Euclidean norm   
equal to 1 and largest component real.   

Arguments   
=========   

JOBVL   (input) CHARACTER*1   
        = 'N': left eigenvectors of A are not computed;   
        = 'V': left eigenvectors of A are computed.   

JOBVR   (input) CHARACTER*1   
        = 'N': right eigenvectors of A are not computed;   
        = 'V': right eigenvectors of A are computed.   

N       (input) INTEGER   
        The order of the matrix A. N >= 0.   

A       (input/output) DOUBLE PRECISION array, dimension (LDA,N)   
        On entry, the N-by-N matrix A.   
        On exit, A has been overwritten.   

LDA     (input) INTEGER   
        The leading dimension of the array A.  LDA >= max(1,N).   

WR      (output) DOUBLE PRECISION array, dimension (N)   
WI      (output) DOUBLE PRECISION array, dimension (N)   
        WR and WI contain the real and imaginary parts,   
        respectively, of the computed eigenvalues.  Complex   
        conjugate pairs of eigenvalues appear consecutively   
        with the eigenvalue having the positive imaginary part   
        first.   

VL      (output) DOUBLE PRECISION array, dimension (LDVL,N)   
        If JOBVL = 'V', the left eigenvectors u(j) are stored one   
        after another in the columns of VL, in the same order   
        as their eigenvalues.   
        If JOBVL = 'N', VL is not referenced.   
        If the j-th eigenvalue is real, then u(j) = VL(:,j),   
        the j-th column of VL.   
        If the j-th and (j+1)-st eigenvalues form a complex   
        conjugate pair, then u(j) = VL(:,j) + i*VL(:,j+1) and   
        u(j+1) = VL(:,j) - i*VL(:,j+1).   

LDVL    (input) INTEGER   
        The leading dimension of the array VL.  LDVL >= 1; if   
        JOBVL = 'V', LDVL >= N.   

VR      (output) DOUBLE PRECISION array, dimension (LDVR,N)   
        If JOBVR = 'V', the right eigenvectors v(j) are stored one   
        after another in the columns of VR, in the same order   
        as their eigenvalues.   
        If JOBVR = 'N', VR is not referenced.   
        If the j-th eigenvalue is real, then v(j) = VR(:,j),   
        the j-th column of VR.   
        If the j-th and (j+1)-st eigenvalues form a complex   
        conjugate pair, then v(j) = VR(:,j) + i*VR(:,j+1) and   
        v(j+1) = VR(:,j) - i*VR(:,j+1).   

LDVR    (input) INTEGER   
        The leading dimension of the array VR.  LDVR >= 1; if   
        JOBVR = 'V', LDVR >= N.   

WORK    (workspace/output) DOUBLE PRECISION array, dimension (LWORK)   
        On exit, if INFO = 0, WORK(1) returns the optimal LWORK.   

LWORK   (input) INTEGER   
        The dimension of the array WORK.  LWORK >= max(1,3*N), and   
        if JOBVL = 'V' or JOBVR = 'V', LWORK >= 4*N.  For good   
        performance, LWORK must generally be larger.   

        If LWORK = -1, then a workspace query is assumed; the routine   
        only calculates the optimal size of the WORK array, returns   
        this value as the first entry of the WORK array, and no error   
        message related to LWORK is issued by XERBLA.   

INFO    (output) INTEGER   
        = 0:  successful exit   
        < 0:  if INFO = -i, the i-th argument had an illegal value.   
        > 0:  if INFO = i, the QR algorithm failed to compute all the   
              eigenvalues, and no eigenvectors have been computed;   
              elements i+1:N of WR and WI contain eigenvalues which   
              have converged.   

=====================================================================   
```

Basically, I'm wondering how the arguments are converted over once the lapack-wrap macro creates scheme bindings. Then I want to test the bindings via `csi` to see if everything worked properly. Thanks again for your help thus far.

**Comment by iraikov** [comment]
Hi Jeremy,

   When you want to wrap a Fortran or C function that takes pointers as
arguments, you have to create SRFI-4 vectors of one element, and pass those
to the function.
For instance, if the argument is of type int\* you must pass it an
s32vector, if it is a double\* you must pass it an f64vector and so on.
You can take a look at the seulex egg, it is a wrapper for a Fortran ODE
solver that follows the same conventions:

http://bugs.call-cc.org/browser/release/4/seulex/trunk/seulex.scm

I will take a look at your code some this week.

  -Ivan

**Comment by ThatGeoGuy** [comment]
Hey Ivan, 

Thanks for the advice. I believe that was the piece I was missing (wrapping things in an s32vector). I will attempt to come up with a reasonable example to test the functions later this week. Afterwards I'll attempt to sort out getting the egg to install properly (beyond my own system), what with requiring -llapack, -lf77blas, and -lgfortran.

**Comment by ThatGeoGuy** [comment]
Hey Ivan, 

I keep getting the error:

```
Error: (dgeev_) matrix A is allocated 9 elements but given dimensions are #s32(3) by #s32(3)
```

I think I may have tracked this down to lines 123 to 132 (perhaps 144?) in the lapack-wrap macro, which are as follows:

```
,(if vsize
   `(,%begin
      (let ((,asize (,vsize a)))
       ,(if (memq 'm fn)
          `(if (< ,asize (fx* m n))
             (atlas-lapack:error ',fname (conc "matrix A is allocated " ,asize " elements "
                                               "but given dimensions are " m " by " n)))
          `(if (< ,asize (fx* n n))
             (atlas-lapack:error ',fname (conc "matrix A is allocated " ,asize " elements "
                                               "but given dimensions are " n " by " n)))))
```

Unfortunately, I am unsure what I can do to fix this issue. I have tried wrapping m and n inside of `(s32vector-ref m 0)`, however, this did not work and ended up with `dgeev_` running into an infinite loop. I believe this may be tied to an issue I was encountering before, where LDA was being set as an optional parameter, and then treated specially within the macro. If I try to run the same test (see tests/dgeev_ at https://github.com/ThatGeoGuy/chicken-lapack/blob/lapack-extras/tests/dgeev_.scm) with LDA set to e.g. (s32vector 4), the exact same error occurs:

```
 Error: (dgeev_) matrix A is allocated 9 elements but given dimensions are #s32(3) by #s32(3)
```

If LDA was used properly, then I suspect it should say that the given dimensions are #s32(3) by #s32(4). Ivan, can you check over the lapack-wrap macro at https://github.com/ThatGeoGuy/chicken-lapack/blob/lapack-extras/lapack-extras.scm#L84-L159 and help me figure out what's going on with the LDA parameter and how to fix the size check for the input matrix A?

EDIT: I have since updated the files to reflect the fact that I reintroduced LDA, LDVL, and LDVR as optional parameters, and solved the issue above regarding the dimension check of A, but am still getting the following error:

```
Error: (f64vector-length) bad argument type - not a structure of the required type #s32(3)
```

I cannot for the life of me figure out why this is popping up, especially given that N should never be referenced by f64vector-length in the first place.

**Comment by iraikov** [comment]
Hi Jeremy,

   Your n dimension seem to be a vector, and it needs to be a scalar.  So
instead of (define N (s32vector 3)) you need to have (define N 3).

 -Ivan

**Comment by ThatGeoGuy** [comment]
Hey Ivan,

I'm not sure I understand. Mostly because the function signature in C as per http://netlib.org/clapack/what/double/dgeev.c specifies that `n` is a pointer to an integer, which is a typedef for a long int, or a 32 bit integer. I looked again at the types, and it appears I am making some odd assumptions, primarily about types. I get the following:

``` C
typedef long int integer;
typedef unsigned long int uinteger;
typedef char *address;
typedef short int shortint;
typedef float real;
typedef double doublereal;
typedef struct { real r, i; } complex;
typedef struct { doublereal r, i; } doublecomplex;
typedef long int logical;
typedef short int shortlogical;
typedef char logical1;
typedef char integer1;
```

Mostly I'm noticing that complex numbers will present a problem. Currently I'm testing `dgeev_`, which uses doubles, so I expect this isn't going to be a huge issue. In any case, changing `(define N (s32vector 3))` to `(define N 3)` does nothing, as I still get an error with respect to `f64vector-length` being called on the integer itself. Does bind / srfi-4 do any kind of checking with respect to the length based on the type it detects for the function signature? I have checked, but I can't for the life of me find where or why it would call `f64vector-length` on the first parameter. I tried testing `sgeev_` as well, and the same thing happened, though with f32vector-length. I know that lapack-wrap refers to the length checker as `,vsize` in the macro, but I don't see any indication of `n` being checked for length. 

EDIT: Interestingly enough, if I don't use the previous definitions: 

``` C
typedef float real;
typedef double doublereal;
typedef float complex;
typedef double doublecomplex;
typedef int integer;
```

I get compiler errors when trying to compile the egg. Strange, but this may be perhaps because it was compiled with ATLAS instead of on its own. Either way, `dgeev_` interfaces a Fortran function, so I would expect to pass a pointer as per your earlier instruction.

**Comment by iraikov** [comment]
Hi Jeremy,

```
I was confused by the code you had pasted above, it seemed that you
```

pass n as a vector but you treat it is a scalar. I see now in your actual
code that you have nval defined as (s32vector n 0) which seems ok. But
nevertheless I think there is confusion about what you pass to the Scheme
function and what you pass to the C function. Even if the C function takes
a pointer to integer, this does not mean that the Scheme function must take
an s32vector. You could have the n argument be a scalar, and then wrap it
in a vector and pass it to the C function. I will look at the code a bit
more and try to figure out why you are getting this error.

  -Ivan

**Comment by ThatGeoGuy** [comment]
Thanks Ivan. If you think there's a more elegant way to wrap this interface, please let me know, I'd be happy to try and refactor the macro. For the time being, however, I have been attempting to get the function to run without error, and then work on making it elegant.

**Comment by iraikov** [comment]
Hi Jeremy,

   I think you will have to make the n argument to the Scheme function be a
scalar, because the calculation of the lda, ldvl, ldvr variables depends on
n being a scalar. Alternatively, you can change the definitions of ld... to
use s32vector-ref on n the way you do with nval. The code as currently
written treats n inconsistently.

   -Ivan

**Comment by ThatGeoGuy** [comment]
Fair point. However, lda, ldvl, and ldvr are also treated as type `integer *` in C. So I need to wrap all of these in s32vectors before passing them to the C function? 

Either way, making the change is easy, however the following error still persists: 

```
Error: (f64vector-length) bad argument type - not a structure of the required type
3
```

Any thoughts on this? I've pushed a new update to the lapack-extras branch, see commit 6c300e61f8f0fd73dd18c79603b67973cdb8440d (https://github.com/ThatGeoGuy/chicken-lapack/commit/6c300e61f8f0fd73dd18c79603b67973cdb8440d).

**Comment by iraikov** [comment]
Hi Jeremy,

   Yes, you will need to make n, lda, ldvl, and ldvr to be s32vectors
before you pass them to the C function.
You can rename the n argument to something like nvec, and create it like
opiv is created in the current code.

   -Ivan

**Comment by iraikov** [comment]
One more thing: it looks like the Scheme procedures and the C procedures
have the same names. You need to make sure that the Scheme procedures don't
have names ending in underscore, otherwise they just end up recursively
calling themselves.

   -Ivan

**Comment by ThatGeoGuy** [comment]
Oh, thanks, that would certainly be helpful. That may be why I had previously run into what I believed to be an infinite loop. I will try to modify some things according to the above suggestions. That said, however, I assume I need to modify cfname to accomodate the underscore suffix. I should be able to do this quite quickly.

**Comment by iraikov** [comment]
Ok, I think I got things to work, patch is attached. Let me know if
anything is still unclear.

On Mon, Mar 16, 2015 at 10:18 PM, Jeremy Steward notifications@github.com
wrote:

> Oh, thanks, that would certainly be helpful. That may be why I had
> previously run into what I believed to be an infinite loop. I will try to
> modify some things according to the above suggestions. That said, however,
> I assume I need to modify cfname to accomodate the underscore suffix. I
> should be able to do this quite quickly.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/iraikov/chicken-lapack/issues/2#issuecomment-82112354
> .

diff --git a/lapack-extras.scm b/lapack-extras.scm
index 708d92d..3050e31 100644
--- a/lapack-extras.scm
+++ b/lapack-extras.scm
@@ -20,18 +20,18 @@

 (module lapack-extras
   (
-   sgeev_
-   dgeev_
-   cgeev_
-   zgeev_
-   sgeev_!
-   dgeev_!
-   cgeev_!
-   zgeev_!
-   unsafe-sgeev_!
-   unsafe-dgeev_!
-   unsafe-cgeev_!
-   unsafe-zgeev_!
-   sgeev
-   dgeev
-   cgeev
-   zgeev
-   sgeev!
-   dgeev!
-   cgeev!
-   zgeev!
-   unsafe-sgeev!
-   unsafe-dgeev!
-   unsafe-cgeev!
-   unsafe-zgeev!
  )
  (import chicken scheme data-structures foreign)
  (use srfi-4 blas bind)
  @@ -84,7 +84,7 @@ EOF
  (define-syntax lapack-wrap
   (lambda (x r c)
     (let\* ((fn      (cadr x))
-           (cfname  (car fn))
-           (cfname  (string->symbol (conc (symbol->string (car fn)) "_")))
          (ret     (caddr x))
          (errs    (cadddr x))
          (vsize   (car (cddddr x)))
  @@ -98,6 +98,7 @@ EOF
                     (if (null? args) (cons fname sig)
                       (let ((x (car args)))
                        (let ((sig (case x
- ```
                                   ((nvec) (cons 'n sig))
                                 ((opiv) sig)
                                 ((lda)  sig)
                                 ((ldb)  sig)
  ```
  
  @@ -118,13 +119,13 @@ EOF
  
     `(,%define ,fsig
        (,%let-optionals rest ,(if (memq 'ldb fn)
-                                  `((lda ,(if (memq 'm fn) 'm 'n))
-                                    (ldb ,(if (memq 'm fn) 'm 'n))
-                                    (ldvl ,(if (memq 'm fn) 'm 'n))
-                                    (ldvr ,(if (memq 'm fn) 'm 'n)))
-                                  `((lda ,(if (memq 'm fn) 'm 'n))
-                                    (ldvl ,(if (memq 'm fn) 'm 'n))
-                                    (ldvr ,(if (memq 'm fn) 'm 'n))))
-                                  `((lda (s32vector ,(if (memq 'm fn) 'm 'n)))
-                                    (ldb (s32vector ,(if (memq 'm fn) 'm 'n)))
-                                    (ldvl (s32vector ,(if (memq 'm fn) 'm 'n)))
-                                    (ldvr (s32vector ,(if (memq 'm fn) 'm 'n))))
-                                  `((lda (s32vector ,(if (memq 'm fn) 'm 'n)))
-                                    (ldvl (s32vector ,(if (memq 'm fn) 'm 'n)))
- ```
                                 (ldvr (s32vector ,(if (memq 'm fn) 'm 'n)))))
  
      ,(if vsize
         `(,%begin
  ```
  
  @@ -155,6 +156,7 @@ EOF
                   (let ((x (car fn)))
                    (let ((bnds (case x
                                  ((opiv)  (cons `(opiv (make-s32vector n)) bnds))
-                                   ((nvec)  (cons `(nvec (s32vector n)) bnds))
                                  (else    (if (and copy (memq x ret))
                                             (cons`(,x (,copy ,x)) bnds)
                                             bnds)))))
  @@ -201,7 +203,7 @@ EOF
       ,ret ,errs (lambda (v) (fx/ 2 (f64vector-length v))) zcopy))))
     )

-(lapack-wrapx (geev_ jobvl jobvr n a lda wr wi vl ldvl vr ldvr work lwork info_)
+(lapack-wrapx (geev jobvl jobvr nvec a lda wr wi vl ldvl vr ldvr work lwork info_)
               (a wr wi vl vr work)
               ((lambda (i) (conc i "-th element had an illegal value."))
                (lambda (i) (conc "QR algorithm failed. Elements "
diff --git a/tests/dgeev_.scm b/tests/dgeev_.scm
index d496dae..2b20ee5 100644
--- a/tests/dgeev_.scm
+++ b/tests/dgeev_.scm
@@ -13,4 +13,5 @@
 (define LWORK (s32vector 12))
 (define INFO (s32vector 0))

-(dgeev_ "V" "V" N A WR WI VL VR WORK LWORK INFO)
+(dgeev "V" "V" N A WR WI VL VR WORK LWORK INFO)
+(print INFO)

**Comment by ThatGeoGuy** [comment]
Hi Ivan, 

My only question is how to apply a patch. I've only done so once, so as far as I know I need to paste the diffs into a patch file and call `patch pfile.p0` or something? I apologize if this sounds like a noob question, I almost never encounter raw patches nowadays.

**Comment by iraikov** [comment]
patch -p1 < lapack-extras.diff should do the job. The --dry-run option to
patch will print the results of the patch without changing the file, useful
to make sure a patch is ok.

  -Ivan

**Comment by ThatGeoGuy** [comment]
Ok, seems the patch is malformed. I get an error on line 7. I tried with patch -p1 and git-apply, and both failed me. 

Could you send again, perhaps as just a file this time?

**Comment by iraikov** [comment]
Oh, maybe Gmail mangled it. Here is a gzipped patch, hopefully this will
work.

**Comment by ThatGeoGuy** [comment]
Hey Ivan, while it appears your patch worked for sgeev and dgeev (I applied it manually), it does not work for cgeev and zgeev. I believe this is for two reasons: 

The first, is that the `lapack-wrapx` macro contains an error for complex and double-complex arguments, and I believe the error persists within the original atlas-lapack egg as well. Specifically the line:

```
       (lapack-wrap ,(cons (string->symbol (conc "c" (symbol->string (car fn)))) (cdr fn))
        ,ret ,errs (lambda (v) (fx/ 2 (f32vector-length v))) ccopy)
```

The statement `(fx/ 2 (f32vector-length v))` should have its arguments in the reverse order, i.e. `(fx/ (f32vector-length v) 2)`. Dividing the length of the vector from 2 makes no sense here, since it will (usually) end up as 0 with fixnum/integer division. I will submit another issue after posting this to reflect the fact that this issue is reflected in the `atlas-lapack` module as well. 

The second problem I suspect is because `cgeev_` and `zgeev_` take different arguments than `dgeev_` and `sgeev_`. Specifically, where `sgeev_` and `dgeev_` take two arguments which return the values of the real (WR) and imaginary (WI) components of the Eigenvectors, `cgeev_` and `zgeev_` only take one value (W). Furthermore, there is an additional parameter before `info` which is called `rwork`. I think it's possible to modify `lapack-wrapx` in place to support the change in arguments; however, it may be simpler and may reduce the complexity of the macro to split lapack-wrapx into two macros, at least within the `lapack-extras` module. 

By splitting the `lapack-wrapx` macro, the return type and arguments of the complex number functions can be differentiated as is the case here. Moreover, for some functions in LAPACK, such as `dsygv_` or `zhegv_` don't have the exact counterparts in the corresponding complex - real space. Hermitian matrices are subtly different than symmetric matrices in the complex space, so there is an API division here as well. I think splitting the `lapack-wrapx` macro into `lapack-wrap-sd` and `lapack-wrap-cz` is worthwhile, as it will reduce the complexity of each individual macro, and will reflect the fact that the API is split within LAPACK itself. I do not know if this was necessary before when dealing with the solvers and inversion functions that ATLAS optimizes (a cursory look would suggest that it doesn't), but when dealing with other driver routines I believe this is an important distinction to make. 

In any case, I will try to work on this when I get some more time available, but for now I'll let this comment serve as documentation for myself when returning to the problem later. My time over the next couple weeks is somewhat limited, but I will try my best to put in a few hours when possible.

**Comment by iraikov** [comment]
Hi Jeremy, you are right, I never actually tested the complex variants of the functions. I agree that the right thing to do is to split the wrapper macros to take care of these cases. I am also a bit short on time at the moment, but I will try to look at this while it is still fresh in my memory. Thanks for raising these issues.

**Comment by ThatGeoGuy** [comment]
I've submitted a PR that should definitively solve this issue. I will start working towards other issues that I had brought up back in March. Apologies for picking this up again so late, but I've only recently finished finals and come upon some free time again.

**Comment by iraikov** [comment]
Thanks, I will look at it. No need to apologize, we are all busy :-)

**Comment by ThatGeoGuy** [comment]
With PR #6 merged, I'll close this issue. Future additions to `lapack-extras` will come, and hopefully with far less effort now that much of the lifting has been done.
